spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java

```diff
@@ -0,0 +1,314 @@
+  private final boolean caseSensitive;
+  private final Schema expectedSchema;
+  private final Long splitSize;
+  private final Integer splitLookback;
+  private final Long splitOpenFileCost;
+  private final boolean localityPreferred;
+  private final InitialOffsetStore initialOffsetStore;
+  private final StreamingOffset initialOffset;
+
+  SparkMicroBatchStream(JavaSparkContext sparkContext,
+                        Table table, boolean caseSensitive, Schema expectedSchema,
+                        CaseInsensitiveStringMap options, String checkpointLocation) {
+    this.sparkContext = sparkContext;
+    this.table = table;
+    this.caseSensitive = caseSensitive;
+    this.expectedSchema = expectedSchema;
+    this.localityPreferred = Spark3Util.isLocalityEnabled(table.io(), table.location(), options);
+    this.splitSize = Optional.ofNullable(Spark3Util.propertyAsLong(options, SparkReadOptions.SPLIT_SIZE, null))
+        .orElseGet(() -> PropertyUtil.propertyAsLong(table.properties(), SPLIT_SIZE, SPLIT_SIZE_DEFAULT));
+    this.splitLookback = Optional.ofNullable(Spark3Util.propertyAsInt(options, SparkReadOptions.LOOKBACK, null))
+        .orElseGet(() -> PropertyUtil.propertyAsInt(table.properties(), SPLIT_LOOKBACK, SPLIT_LOOKBACK_DEFAULT));
+    this.splitOpenFileCost = Optional.ofNullable(
+        Spark3Util.propertyAsLong(options, SparkReadOptions.FILE_OPEN_COST, null))
+        .orElseGet(() -> PropertyUtil.propertyAsLong(table.properties(), SPLIT_OPEN_FILE_COST,
+            SPLIT_OPEN_FILE_COST_DEFAULT));
+    this.initialOffsetStore = InitialOffsetStore.getInstance(table.io(), checkpointLocation);
+    this.initialOffset = getOrWriteInitialOffset(table, initialOffsetStore);
+  }
+
+  @Override
+  public Offset latestOffset() {
+    table.refresh();
+    Snapshot latestSnapshot = table.currentSnapshot();
+    if (latestSnapshot == null) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    String addedFilesValue = latestSnapshot.summary().get(SnapshotSummary.ADDED_FILES_PROP);
+    long addedFiles = addedFilesValue == null ?
+        Iterables.size(latestSnapshot.addedFiles()) :
+        Long.parseLong(addedFilesValue);
+
+    // a readStream on an Iceberg table can be started from 2 types of snapshots
+    // 1. a valid starting Snapshot:
+    //      when this valid starting Snapshot is the initialOffset - then, scanAllFiles must be set to true;
+    //      for all StreamingOffsets following this - scanAllFiles must be set to false
+    // 2. START_OFFSET:
+    //      if the stream started on the table from START_OFFSET - it implies - that all the subsequent Snapshots added
+    //      will have all files as net New manifests & hence scanAllFiles can be false.
+    boolean scanAllFiles = !StreamingOffset.START_OFFSET.equals(initialOffset) &&
+        latestSnapshot.snapshotId() == initialOffset.snapshotId();
+    return new StreamingOffset(
+        latestSnapshot.snapshotId(),
+        addedFiles,
```

Comment:
When starting from an existing snapshot (`initialOffset` is not `START_OFFSET`), the end offset for the initial snapshot is the total number of data files in the table and `scanAllFiles` should be true.

The logic here will produce a `latestOffset` that uses the number of files added in that starting snapshot, even though `scanAllFiles` is true. I don't think that's correct.

For example, if I have a table with the following snapshot list:
id | added_files | existing files
-- | -- | --
1 | A.parquet, B.parquet | X.parquet, Y.parquet, Z.parquet

The initial offset should be `(1, 0, true)` and the largest offset in snapshot 1 is `(1, 5, true)`. But the latest offset with those starting conditions will be `(1, 2, true)`. The remaining files in that snapshot wouldn't be processed until the next snapshot is added to the table, if I understand correctly. That's a problem because we want to process the snapshot for `scanAllFiles` as quickly as possible since it may be large.

Reply:
ah. nice catch.

Reply:
fixed this.